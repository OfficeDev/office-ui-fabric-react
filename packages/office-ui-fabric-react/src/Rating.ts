export * from './packages/react-inputs/Rating';
