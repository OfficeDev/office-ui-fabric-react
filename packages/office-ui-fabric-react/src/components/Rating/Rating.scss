@import '../../common/common';

$Rating-defaultColor: $ms-color-neutralTertiary;
$Rating-selectedColor: $ms-color-black;
$Rating-hoverColor: $ms-color-themeDarkAlt;
$Rating-disabledColor: $ms-color-neutralTertiaryAlt;

$Rating-smallIconSize: 16px;
$Rating-largeIconSize: 20px;

.ratingStar {
  display: inline-block;
  position: relative;
}

.ratingStarBack {
  color: $Rating-defaultColor;
  width: 100%;

  &.ratingStarDisabled {
    color: $Rating-disabledColor;
     @media screen and (-ms-high-contrast: active) {
      color: $Rating-disabledColor;
    }
  }
}

.ratingStarFront {
  position: absolute;
  height: 100%;
  left: 0;
  top: 0;
  text-align: center;
  vertical-align: middle;
  overflow: hidden;
  color: $Rating-selectedColor;
   @media screen and (-ms-high-contrast: active) {
      color: rgb(255, 255, 0);
   }
}

.ratingButton {
  background: none;
  margin: 0;
  padding: 3px 3px 0px 0px;
  border: none;
}

.rootIsSmall  {
    font-size: $Rating-smallIconSize;
    line-height: $Rating-smallIconSize;
  }

.rootIsLarge  {
    font-size: $Rating-largeIconSize;
    line-height: $Rating-largeIconSize;
}

.labelText {
  @include ms-screenReaderOnly();
}

.ratingFocusZone {
  display: inline-block
}