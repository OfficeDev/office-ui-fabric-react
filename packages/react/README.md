# @uifabric/react

**React components for [Office UI Fabric React](https://dev.microsoft.com/fabric)**

These are not production-ready components and **should never be used in product**. This space is useful for testing new components whose APIs might change before final release.

To import React components:

```js
import { ComponentName } from '@uifabric/react';
```

Once the React component graduates to a production release, the component will be available at:

```js
import { ComponentName } from 'office-ui-fabric-react';
```
