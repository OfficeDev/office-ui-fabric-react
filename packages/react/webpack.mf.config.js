const path = require('path');
const fs = require('fs');
const webpack = require('webpack');
const getResolveAlias = require('@fluentui/scripts/webpack/getResolveAlias');
const { createServeConfig } = require('@fluentui/scripts/webpack/webpack-resources');
const { webpackMerge } = require('just-scripts');

const shared = {
  react: { singleton: true },
  'react-dom': { singleton: true },
};

const rootComponents = fs
  .readdirSync(path.join(__dirname, 'src'))
  .filter(filename => filename[0].toUpperCase() === filename[0])
  .map(filename => filename.replace(/\.tsx?/, ''));
const rootComponentsExposes = {};
for (const component of rootComponents) {
  rootComponentsExposes[`./lib/${component}`] = `@fluentui/react/src/${component}`;
}

// TODO: fork-ts-checker-webpack-plugin needs to be taught about the resolve alias -> pathmapping in order for the type checking to work
process.env.SKIP_TYPECHECK = true;

const myConfig = {
  output: {},
  entry: './empty-entrypoint.js',
  mode: 'development',
  devtool: 'cheap-module-source-map',
  optimization: {
    concatenateModules: false,
  },
  module: {
    rules: [],
  },
  plugins: [
    new webpack.container.ModuleFederationPlugin({
      name: 'fluentuiReact',
      filename: 'remoteEntry.js',
      exposes: {
        ...rootComponentsExposes,
        './lib/compat/Button': '@fluentui/react/src/compat/Button',
        '.': '@fluentui/react/src/index',
        './lib/index': '@fluentui/react/src/index',
      },
      shared,
    }),
  ],
  resolve: {
    alias: getResolveAlias(),
  },
  devServer: {
    port: 2345,
    static: './dist',
  },
};

const metaConfig = webpackMerge.merge(createServeConfig(), myConfig);

module.exports = metaConfig;
