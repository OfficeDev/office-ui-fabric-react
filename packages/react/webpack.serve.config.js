const path = require('path');
const resources = require('@uifabric/build/webpack/webpack-resources');

module.exports = resources.createServeConfig({
  entry: './src/demo/index.tsx',

  output: {
    filename: 'demo-app.js',
  },

  externals: {
    react: 'React',
    'react-dom': 'ReactDOM',
  },

  resolve: {
    alias: {
      '@fluentui/react/src': path.join(__dirname, 'src'),
      '@fluentui/react/lib': path.join(__dirname, 'lib'),
      '@fluentui/react': path.join(__dirname, 'lib'),
      'Props.ts.js': 'Props',
      'Example.tsx.js': 'Example',
    },
  },
});
