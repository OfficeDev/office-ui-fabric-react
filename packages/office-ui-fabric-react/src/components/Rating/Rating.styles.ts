import {
  getFocusStyle,
  hiddenContentStyle,
  HighContrastSelector,
  getGlobalClassNames,
} from '../../Styling';
import { IRatingStyleProps, IRatingStyles } from './Rating.types';

const GlobalClassNames = {
  ratingStar: 'ms-RatingStar-container',
  ratingStarBack: 'ms-RatingStar-back',
  ratingStarFront: 'ms-RatingStar-front',
  ratingButton: 'ms-Rating-button',
  rootIsSmall: 'ms-Rating--small',
  rootIsLarge: 'ms-Rating--large',
  labelText: 'ms-Rating-labelText',
  ratingFocusZone: 'ms-Rating-focuszone',
};

function _getColorWithHighContrast(color: string, highContrastColor: string) {
  return {
    color: color,
    selectors: {
      [HighContrastSelector]: {
        color: highContrastColor
      }
    }
  };
}

export function getStyles(props: IRatingStyleProps): IRatingStyles {
  const {
    disabled,
    readOnly,
    theme
  } = props;

  const {
    semanticColors,
    palette
  } = theme;

  const classNames = getGlobalClassNames(GlobalClassNames, theme);

  const ratingSmallIconSize = '16px';
  const ratingLargeIconSize = '20px';

  const ratingStarUncheckedColor = palette.neutralTertiary;
  const ratingStarCheckedColor = semanticColors.bodyTextChecked;
  const ratingStarDisabledColor = semanticColors.disabledBodyText;

  return {
    root: [
      !disabled && !readOnly && {
        selectors: {
          // This is part 1 of highlighting all stars up to the one the user is hovering over
          '&:hover': {
            selectors: {
              '.ms-RatingStar-back': _getColorWithHighContrast(ratingStarCheckedColor, 'Highlight'),
            }
          }
        }
      }
    ],
    ratingStar: [
      classNames.ratingStar,
      {
        display: 'inline-block',
        position: 'relative'
      }
    ],
    ratingStarBack: [
      classNames.ratingStarBack,
      {
        // TODO: Use a proper semantic color for this
        color: ratingStarUncheckedColor,
        width: '100%'
      },
      disabled && _getColorWithHighContrast(ratingStarDisabledColor, 'GrayText')
    ],
    ratingStarFront: [
      classNames.ratingStarFront,
      {
        position: 'absolute',
        height: '100 %',
        left: '0',
        top: '0',
        textAlign: 'center',
        verticalAlign: 'middle',
        overflow: 'hidden',
      },
      _getColorWithHighContrast(ratingStarCheckedColor, 'Highlight')
    ],
    ratingButton: [
      getFocusStyle(theme, 0),
      classNames.ratingButton,
      {
        backgroundColor: 'transparent',
        padding: '3px 3px 0px 0px',
        margin: '0px',
        border: 'none',
        cursor: 'pointer',
        selectors: {
          '&:disabled': {
            cursor: 'default'
          },
          '&[disabled]': {
            cursor: 'default'
          }
        }
      },
      !disabled && !readOnly && {
        selectors: {
          // This is part 2 of highlighting all stars up to the one the user is hovering over
          '&:hover ~ .ms-Rating-button': {
            selectors: {
              '.ms-RatingStar-back': _getColorWithHighContrast(ratingStarUncheckedColor, 'WindowText'),
              '.ms-RatingStar-front': _getColorWithHighContrast(ratingStarUncheckedColor, 'WindowText'),
            }
          }
        }
      },
      disabled && {
        cursor: 'default'
      },
    ],
    rootIsSmall: [
      classNames.rootIsSmall,
      {
        fontSize: ratingSmallIconSize,
        lineHeight: ratingSmallIconSize
      }
    ],
    rootIsLarge: [
      classNames.rootIsLarge,
      {
        fontSize: ratingLargeIconSize,
        lineHeight: ratingLargeIconSize
      }
    ],
    labelText: [
      classNames.labelText,
      hiddenContentStyle
    ],
    ratingFocusZone: [
      classNames.ratingFocusZone,
      {
        display: 'inline-block',
        paddingBottom: '1px'
      }
    ]
  };
}
