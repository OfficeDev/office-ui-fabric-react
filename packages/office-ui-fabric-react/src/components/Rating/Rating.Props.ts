import * as React from 'react';

export interface IRating {

}

/**
 * Rating component props.
 */
export interface IRatingProps extends React.HTMLProps<HTMLElement> {
  /**
   * Optional callback to access the IRating interface. Use this instead of ref for accessing
   * the public methods and properties of the component.
   */
  componentRef?: (component: IRating) => void;

  /**
   * Selected rating, has to be an integer between min and max
   * If aggregate is not set to true, value is floored
   */
  rating?: number;

  /**
   * Minimum rating, defaults to 1, has to be >= 0
   */
  min?: number;

  /**
   * Maximum rating, defaults to 5, has to be >= min
   */
  max?: number;

  /**
   * Custom icon, defaults to FavoriteStar
   */
  icon?: string;

  /**
   * Size of rating, defaults to small
   */
  size?: RatingSize;

  /**
   * Callback when the rating has changed
   */
  onChanged?: (rating: number) => void;

  /**
   * Optional label for star ratings, will be read by screen readers, defaults to 'Star'.
   */
  ariaLabelIcon?: string;

  /**
   * Optional id of label describing this instance of Rating
   */
  ariaLabelId?: string;

  /**
   * Optional setting to represent aggregate rating values on rating control.
   * If set, rating control will be readonly rating control with partial star support.
   * Also if set, this overrides the disabled prop
   */
  aggregate?: boolean;
}

export enum RatingSize {
  Small = 0,
  Large = 1
}