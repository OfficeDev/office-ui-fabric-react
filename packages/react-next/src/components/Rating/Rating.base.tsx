import * as React from 'react';
import { classNamesFunction, css, format, divProperties, getNativeProps } from '../../Utilities';
import { IProcessedStyleSet } from '../../Styling';
import { Icon } from '../../Icon';
import { FocusZone, FocusZoneDirection, IFocusZoneProps } from '../../FocusZone';
import { IRatingProps, RatingSize, IRatingStyleProps, IRatingStyles } from './Rating.types';
import { useId, useConst, useWarnings } from '@uifabric/react-hooks';

const getClassNames = classNamesFunction<IRatingStyleProps, IRatingStyles>();

interface IRatingStarProps extends React.AllHTMLAttributes<HTMLElement> {
  fillPercentage: number;
  disabled?: boolean;
  readOnly?: boolean;
  classNames: IProcessedStyleSet<IRatingStyles>;
  icon?: string;
}

const RatingStar = (props: IRatingStarProps) => {
  const icon = props.icon || 'FavoriteStarFill';
  return (
    <div className={props.classNames.ratingStar} key={props.id}>
      <Icon className={props.classNames.ratingStarBack} iconName={icon} />
      {!props.disabled && (
        <Icon
          className={props.classNames.ratingStarFront}
          iconName={icon}
          style={{ width: props.fillPercentage + '%' }}
        />
      )}
    </div>
  );
};

export const RatingBase = React.forwardRef<HTMLDivElement, IRatingProps>((props, ref) => {
  const id = useId('Rating');
  const labelId = useId('RatingLabel');
  const {
    ariaLabelFormat,
    allowZeroStars,
    disabled,
    getAriaLabel,
    onChange,
    // eslint-disable-next-line deprecation/deprecation
    onChanged,
    styles,
    // eslint-disable-next-line deprecation/deprecation
    min = 1,
    max = 5,
    readOnly,
    size,
    theme,
    icon = 'FavoriteStarFill',
    unselectedIcon = 'FavoriteStar',
  } = props;

  const minimumStars = useConst<number>(allowZeroStars ? 0 : 1);

  const divProps = getNativeProps<React.HTMLAttributes<HTMLDivElement>>(props, divProperties);

  const classNames = getClassNames(styles!, {
    disabled,
    readOnly,
    theme: theme!,
  });

  const stars = [];

  const starIds = [];

  const getClampedRating = (ratingProp: number): number => {
    return Math.min(Math.max(ratingProp, minimumStars as number), max as number);
  };

  const getInitialValue = (): number | undefined => {
    if (typeof props.rating === 'undefined') {
      return min;
    }

    if (props.rating === null) {
      return undefined;
    }

    return getClampedRating(props.rating);
  };

  const [rating, setRating] = React.useState<number | null | undefined>(getInitialValue());

  const getRating = (): number => {
    if (props.rating !== undefined) {
      return getClampedRating(props.rating);
    }
    if (rating !== undefined && rating !== null) {
      return getClampedRating(rating);
    }
    return 0;
  };

  useDebugWarnings(props);

  const ariaLabel = getAriaLabel ? getAriaLabel(getRating() ? getRating() : 0, max as number) : undefined;

  // When in read-only mode, we allow focus (per ARIA standards) and set up ARIA attributes to indicate element
  // is read-only. https://www.w3.org/TR/wai-aria-1.1/#aria-readonly
  const readOnlyProps: IFocusZoneProps | undefined = readOnly
    ? ({
        allowFocusRoot: true,
        disabled: true,
        'aria-label': ariaLabel,
        'aria-readonly': true,
        'data-is-focusable': true,
        tabIndex: 0,
      } as IFocusZoneProps)
    : undefined;

  const getStarId = (index: number): string => {
    return id + '-star-' + index;
  };

  const onFocus = (value: number, ev: React.FocusEvent<HTMLElement>): void => {
    // const currentRating = getRating();
    if (Math.ceil(rating!) !== value) {
      setRating(value);
      onChange?.(ev, value);
      onChanged?.(value);
    }
  };

  const getLabel = (ratingProp: number): JSX.Element => {
    const text = ariaLabelFormat || '';

    return (
      <span id={`${labelId}-${ratingProp}`} className={classNames.labelText}>
        {format(text, ratingProp, max)}
      </span>
    );
  };

  const getFillingPercentage = (starPosition: number): number => {
    const currentRating = getRating();
    const ceilValue = Math.ceil(currentRating as number);
    let fillPercentage = 100;

    if (starPosition === currentRating) {
      fillPercentage = 100;
    } else if (starPosition === ceilValue) {
      fillPercentage = 100 * ((currentRating as number) % 1);
    } else if (starPosition > ceilValue) {
      fillPercentage = 0;
    }
    return fillPercentage;
  };

  for (let i = minimumStars as number; i <= (max as number); i++) {
    if (i !== 0) {
      const currentRating = getRating();
      const fillPercentage = getFillingPercentage(i);
      const ratingStarProps: IRatingStarProps = {
        fillPercentage,
        disabled,
        classNames: classNames,
        icon: fillPercentage > 0 ? icon : unselectedIcon,
      };

      starIds.push(getStarId(i - 1));

      stars.push(
        <button
          className={css(classNames.ratingButton, {
            [classNames.ratingStarIsLarge]: size === RatingSize.Large,
            [classNames.ratingStarIsSmall]: size !== RatingSize.Large,
          })}
          id={starIds[i - 1]}
          key={i}
          {...(i === Math.ceil(currentRating) ? { 'data-is-current': true } : {})}
          onFocus={onFocus.bind(currentRating, i)}
          onClick={onFocus.bind(currentRating, i)} // For Safari & Firefox on OSX
          disabled={disabled || readOnly ? true : false}
          role="presentation"
          type="button"
        >
          {getLabel(i)}
          <RatingStar key={i + 'rating'} {...ratingStarProps} />
        </button>,
      );
    }
  }

  return (
    <div
      ref={ref}
      className={css('ms-Rating-star', classNames.root, {
        [classNames.rootIsLarge]: size === RatingSize.Large,
        [classNames.rootIsSmall]: size !== RatingSize.Large,
      })}
      aria-label={!readOnly ? ariaLabel : ''}
      id={id}
      {...divProps}
    >
      <FocusZone
        direction={FocusZoneDirection.horizontal}
        className={css(classNames.ratingFocusZone, {
          [classNames.rootIsLarge]: size === RatingSize.Large,
          [classNames.rootIsSmall]: size !== RatingSize.Large,
        })}
        defaultActiveElement={
          getRating() ? starIds[Math.ceil(getRating()) - 1] && '#' + starIds[Math.ceil(getRating()) - 1] : undefined
        }
        {...readOnlyProps}
      >
        {stars}
      </FocusZone>
    </div>
  );
});

function useDebugWarnings(props: IRatingProps) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line react-hooks/rules-of-hooks -- build-time conditional
    useWarnings({
      name: 'Rating',
      props,
      deprecations: { onChanged: 'onChange' },
    });
  }
}
