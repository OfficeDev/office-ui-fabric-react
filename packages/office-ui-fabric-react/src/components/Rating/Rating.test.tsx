/* tslint:disable:no-unused-variable */
import * as React from 'react';
/* tslint:enable:no-unused-variable */

import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';

import { Rating } from './Rating';

describe('Rating', () => {
  it('Can change rating.', () => {
    let exception;
    let threwException = false;
    let rating;
    try {
      rating = ReactTestUtils.renderIntoDocument<Rating>(
        <Rating
          rating={ 2 }
        />
      );
    } catch (e) {
      exception = e;
      threwException = true;
    }
    expect(threwException).toEqual(false);
    let renderedDOM = ReactDOM.findDOMNode(rating as React.ReactInstance);

    let ratingButtons = renderedDOM.querySelectorAll('.ms-Rating-button');
    let ratingFrontStars = renderedDOM.querySelectorAll('.ms-RatingStar-front');

    const checkState = (ratingToCheck: number, state: string) => {
      let iconElement = ratingFrontStars[ratingToCheck - 1] as HTMLElement;
      let width = iconElement.style.width;
      expect(width).to.be.eq(state, 'Rating star width should be ${state}');

    };

    checkState(1, '100%');
    checkState(2, '100%');
    checkState(3, '0%');
    checkState(4, '0%');
    checkState(5, '0%');

    ReactTestUtils.Simulate.focus(ratingButtons[0]);

    checkState(1, '100%');
    checkState(2, '0%');
    checkState(3, '0%');
    checkState(4, '0%');
    checkState(5, '0%');
  });

  it('Clamps input rating to allowed range.', () => {
    let exception;
    let threwException = false;
    let rating;
    try {
      rating = ReactTestUtils.renderIntoDocument<Rating>(
        <Rating
          rating={ 10 }
        />
      );
    } catch (e) {
      exception = e;
      threwException = true;
    }
    expect(threwException).toEqual(false);

    let renderedDOM = ReactDOM.findDOMNode(rating as React.ReactInstance);

    let ratingFrontStars = renderedDOM.querySelectorAll('.ms-RatingStar-front');

    const checkState = (ratingToCheck: number, state: string) => {
      let iconElement = ratingFrontStars[ratingToCheck - 1] as HTMLElement;
      let width = iconElement.style.width;
      expect(width).to.be.eq(state, 'Rating star width should be ${state}');
    };

    expect(ratingFrontStars.length).to.be.eq(5, 'Number of rating stars are clamped to 5');
    checkState(1, '100%');
    checkState(2, '100%');
    checkState(3, '100%');
    checkState(4, '100%');
    checkState(5, '100%');

  });

  it('Clamps input rating to nearest number when halfstar is not enabled.', () => {
    let exception;
    let threwException = false;
    let rating;
    try {
      rating = ReactTestUtils.renderIntoDocument<Rating>(
        <Rating
          rating={ 2.5 }
        />
      );
    } catch (e) {
      exception = e;
      threwException = true;
    }
    expect(threwException).to.be.false;

    let renderedDOM = ReactDOM.findDOMNode(rating as React.ReactInstance);
    let ratingFrontStars = renderedDOM.querySelectorAll('.ms-RatingStar-front');

    const checkState = (ratingToCheck: number, state: string) => {
      let iconElement = ratingFrontStars[ratingToCheck - 1] as HTMLElement;
      let width = iconElement.style.width;
      expect(width).to.be.eq(state, 'Rating star width should be ${state}');
    };

    checkState(1, '100%');
    checkState(2, '100%');
    checkState(3, '0%');
    checkState(4, '0%');
    checkState(5, '0%');
  });

  it('Half star is displayed when it is enabled.', () => {
    let exception;
    let threwException = false;
    let rating;
    try {
      rating = ReactTestUtils.renderIntoDocument<Rating>(
        <Rating
          rating={ 2.5 }
          enableHalfStar={ true }
        />
      );
    } catch (e) {
      exception = e;
      threwException = true;
    }
    expect(threwException).to.be.false;

    let renderedDOM = ReactDOM.findDOMNode(rating as React.ReactInstance);
    let ratingFrontStars = renderedDOM.querySelectorAll('.ms-RatingStar-front');

    const checkState = (ratingToCheck: number, state: string) => {
      let iconElement = ratingFrontStars[ratingToCheck - 1] as HTMLElement;
      let width = iconElement.style.width;
      expect(width).to.be.eq(state, 'Rating star width should be ${state}');
    };

    checkState(1, '100%');
    checkState(2, '100%');
    checkState(3, '50%');
    checkState(4, '0%');
    checkState(5, '0%');

  });

  it('When rating is disabled cannot change rating', () => {
    let exception;
    let threwException = false;
    let choiceGroup;
    try {
      choiceGroup = ReactTestUtils.renderIntoDocument<Rating>(
        <Rating
          disabled={ true }
        />
      );
    } catch (e) {
      exception = e;
      threwException = true;
    }
    expect(threwException).toEqual(false);

    let renderedDOM = ReactDOM.findDOMNode(choiceGroup as React.ReactInstance);
    let ratingButtons = renderedDOM.querySelectorAll('.ms-Rating-button');
    expect((ratingButtons[0] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[1] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[2] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[3] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[4] as HTMLButtonElement).disabled).to.be.eq(true);

  });

  it('When rating is readonly cannot change rating', () => {
    let exception;
    let threwException = false;
    let choiceGroup;
    try {
      choiceGroup = ReactTestUtils.renderIntoDocument<Rating>(
        <Rating
          readOnly={ true }
          rating={ 2 }
        />
      );
    } catch (e) {
      exception = e;
      threwException = true;

    }
    expect(threwException).to.be.false;

    let renderedDOM = ReactDOM.findDOMNode(choiceGroup as React.ReactInstance);
    let ratingButtons = renderedDOM.querySelectorAll('.ms-Rating-button');
    expect((ratingButtons[0] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[1] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[2] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[3] as HTMLButtonElement).disabled).to.be.eq(true);
    expect((ratingButtons[4] as HTMLButtonElement).disabled).to.be.eq(true);

    let ratingFrontStars = renderedDOM.querySelectorAll('.ms-RatingStar-front');
    const checkState = (ratingToCheck: number, state: string) => {
      let iconElement = ratingFrontStars[ratingToCheck - 1] as HTMLElement;
      let width = iconElement.style.width;
      expect(width).to.be.eq(state, 'Rating star width should be ${state}');
    };

    checkState(1, '100%');
    checkState(2, '100%');
    checkState(3, '0%');
    checkState(4, '0%');
    checkState(5, '0%');
  });

});
