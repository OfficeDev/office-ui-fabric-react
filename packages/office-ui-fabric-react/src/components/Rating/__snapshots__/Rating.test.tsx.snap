// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating Renders Rating correctly 1`] = `
<div
  aria-label=""
  className="ms-Rating-star"
>
  <div
    aria-describedby={undefined}
    aria-labelledby={undefined}
    className="ms-FocusZone ms-Rating-focuszone"
    data-focuszone-id="FocusZone2"
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDownCapture={[Function]}
    role="presentation"
    tabIndex={-1}
  >
    <button
      className="ms-Rating-button ms-Rating--small undefined"
      data-is-current={true}
      disabled={false}
      onFocus={[Function]}
      role="presentation"
    >
      <span
        className="ms-Rating-labelText"
        id="RatingLabel1-1"
      >
        1 of 5 stars selected
      </span>
      <div
        className="ms-RatingStar-container"
      >
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-back
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
        />
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-front
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
    </button>
    <button
      className="ms-Rating-button ms-Rating--small undefined"
      disabled={false}
      onFocus={[Function]}
      role="presentation"
    >
      <span
        className="ms-Rating-labelText"
        id="RatingLabel1-2"
      >
        2 of 5 stars selected
      </span>
      <div
        className="ms-RatingStar-container"
      >
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-back
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
        />
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-front
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
          style={
            Object {
              "width": "0%",
            }
          }
        />
      </div>
    </button>
    <button
      className="ms-Rating-button ms-Rating--small undefined"
      disabled={false}
      onFocus={[Function]}
      role="presentation"
    >
      <span
        className="ms-Rating-labelText"
        id="RatingLabel1-3"
      >
        3 of 5 stars selected
      </span>
      <div
        className="ms-RatingStar-container"
      >
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-back
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
        />
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-front
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
          style={
            Object {
              "width": "0%",
            }
          }
        />
      </div>
    </button>
    <button
      className="ms-Rating-button ms-Rating--small undefined"
      disabled={false}
      onFocus={[Function]}
      role="presentation"
    >
      <span
        className="ms-Rating-labelText"
        id="RatingLabel1-4"
      >
        4 of 5 stars selected
      </span>
      <div
        className="ms-RatingStar-container"
      >
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-back
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
        />
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-front
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
          style={
            Object {
              "width": "0%",
            }
          }
        />
      </div>
    </button>
    <button
      className="ms-Rating-button ms-Rating--small undefined"
      disabled={false}
      onFocus={[Function]}
      role="presentation"
    >
      <span
        className="ms-Rating-labelText"
        id="RatingLabel1-5"
      >
        5 of 5 stars selected
      </span>
      <div
        className="ms-RatingStar-container"
      >
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-back
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
        />
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-RatingStar-front
              {
                display: inline-block;
              }
          data-icon-name="FavoriteStarFill"
          role="presentation"
          style={
            Object {
              "width": "0%",
            }
          }
        />
      </div>
    </button>
  </div>
</div>
`;
